mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -1740,3 +1741,294 @@ Value mlir::rock::addPassThroughIndices(OpBuilder &b, Value transformed,
   }
   return ret;
 }
+
+enum DimType { Upper = 0, Lower = 1 };
+
+struct TransformAttrArgs {
+  rock::TransformType type;
+  std::pair<SmallVector<StringRef>, SmallVector<StringRef>> perservedNames;
+  std::pair<SmallVector<uint32_t>, SmallVector<uint32_t>> preservedDims;
+  SmallVector<int64_t> params;
+};
+
+template <DimType Type>
+SmallVector<uint32_t>
+getPerservedIndices(rock::TransformAttr tr,
+                    const SetVector<int64_t> &globalRemoveIndicesSet) {
+  SmallVector<uint32_t> preservedIndices;
+  auto upperDims =
+      Type == DimType::Upper ? tr.getUpperDims() : tr.getLowerDims();
+  for (auto dim : upperDims) {
+    if (!globalRemoveIndicesSet.contains(dim)) {
+      preservedIndices.push_back(dim);
+    }
+  }
+  return preservedIndices;
+}
+
+template <DimType Type>
+void propagatePreservedNames(rock::TransformAttr tr, TransformAttrArgs &args) {
+  const auto &preservedDims = std::get<Type>(args.preservedDims);
+  auto &preseredNames = std::get<Type>(args.perservedNames);
+  auto names = Type == DimType::Upper ? tr.getUpperNames() : tr.getLowerNames();
+  auto dims = Type == DimType::Upper ? tr.getUpperDims() : tr.getLowerDims();
+  assert(names.size() == dims.size());
+  for (auto [idx, dim] : llvm::enumerate(dims)) {
+    if (llvm::is_contained(preservedDims, dim)) {
+      preseredNames.push_back(names[idx]);
+    }
+  }
+}
+
+template <DimType Type>
+SmallVector<uint32_t> getDifference(rock::TransformAttr tr,
```

Comment:
Maybe this is Rust brain talking, but why does this function take the entire struct and then pick a half of it to operate on based on a template parameter? You could just pass references to the sections you need?

Reply:
(redundant with above general structure notes)

Reply:
My comment is below.